# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def addTwo(x):
    return x + 2
print(addTwo(4))

#Lambda Function Syntax
#Example of calling a Lambda function (without a variable)
print((lambda x: x + 2)(4))


6
6


#### Saving to a Variable

In [4]:
f_test = lambda x:x + 2
print(f_test(6))

8


#### Multiple Inputs

In [7]:
print((lambda x,y,z:x * y * z)(3,5,8))
#using a variable
x_test = lambda x,y,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [17]:
def multiply(f,num):
    #f expects a lambda function
    #num expects a number
    return (num)
multiply(lambda x: x*x, 4)



4

#### Returning a Lambda from a Function

In [15]:
# regular function
def multiply_text(num):
    return num * 4
#function within a function
def returnFunc():
    test = 4          # not doing anything
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

#Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c

r_lamb = returnLam(4,5)
print(r_lamb)
print(r_lamb(5,5))    #5,5 = x,a (from line 18)
#or
l_lamb = returnLam(4,5)(5,5)
print(l_lamb)

<function returnFunc.<locals>.multiply at 0x000001C3188DD430>
8
<function returnLam.<locals>.<lambda> at 0x000001C3188DD5E0>
19
19


#### If Statements within Lambdas

In [19]:
#Lambda x : True if (condition) else False
f_condition = lambda num: num * 2 if num > 10 else num + 2
print(f_condition(8))
print(f_condition(12))
print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [28]:
x = lambda num: num**3
print(x(4))


64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [34]:
# map(function, iterable(list, dict, tuple, set))
# normal use is with pre-defined function - or lambda
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num,num2

numbers = [4,11,20,15,20,11,42]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared,numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (15, 2), (20, 6)]


#### Using Lambda's with Map

In [36]:
# map(lambda x: x + 2, list)
# using a lambda in a map condenses code
#list(map(lambda x:x + 2, list))
squared_nums_lamb = list(map(lambda x,y: (x**2,y**2) if x<10 and y<10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)

[(16, 16), (11, 10), (20, 3), (15, 2), (20, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [39]:
numbers = [4,11,20,15,20,11,42]
more_nums = [4,10,3,2,6]

x = list(map(lambda x,y: (x*2-1, y*2-1), numbers,more_nums))
print(x)

[(7, 7), (21, 19), (39, 5), (29, 3), (39, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [44]:
names = ['Bob', 'Andy','Max','Evan','Angelica']
def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names,names))
print(new_names_list)
print(a_names)

['Andy', 'Angelica']
<function a_names at 0x000001C3188DDAF0>


#### Using Lambda's with Filter()

In [46]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [50]:
from statistics import mean

nums = [2,7,4,2,1,6,9,4,4,4,9]
x = list(filter(lambda num: True if num < mean(nums) else False, nums))
print(x)

[2, 4, 2, 1, 4, 4, 4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [51]:
from functools import reduce

# reduce(function, iterable)

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1+num2
result_add = reduce(addNums,list_1)
print(result_add)

30


#### Using Lambda's with Reduce()

In [52]:
result_lamb = reduce(lambda x,y: x+y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [54]:
my_list = [1,2,3,4,5]
x = reduce(lambda x,y: x*y, my_list)
print(x)

120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [56]:
def addNums(num):
    #set base case for recursive function
    if num <= 1:
        print('addNums(1) = 1 (base case)')
        return num
    else:
        print(f"addNums({num}) = (num) + addNums ({num-1})")
        return num + addNums(num-1)

addNums(5)    

addNums(5) = (num) + addNums (4)
addNums(4) = (num) + addNums (3)
addNums(3) = (num) + addNums (2)
addNums(2) = (num) + addNums (1)
addNums(1) = 1 (base case)


15

#### Writing a Factorial Function

In [58]:
def factorial(num):
    if num <=1:
        return num
    else:
        return num * factorial(num-1)
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [64]:
def subtract(num):
    if num <=1:
        return 1
    else:
        return num - subtract(num-1)
subtract(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [69]:
def my_range(stop,start,step = 2):
    while start < stop:
        yield start
        start += step
        
for i in my_range(20,2):
    my_generator_value = i
    print(my_generator_value)

my_range(20,2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x000001C318925F20>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [84]:
def squareGen(num):
    while num > 0:
        yield num**2
        num -= 1
        
for i in squareGen(6):
    sq = i
    print(sq)


36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [97]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

placesOnly = list(filter(lambda place: True if place.strip() != ""  else False, places))
print(placesOnly)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [214]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

def sortNames(author):
    author.sort(key = lambda x: x.lower().split()[-1])
    return author

print(sortNames(author))


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [186]:
celsius = list(map(lambda x :(x[0] , (9/5)*x[1] + 32),places))
print(celsius)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [164]:
def fibseq(num):
    if num == 0:
        return 1
    elif num == 1:
        return 1
    else:
        return fibseq(num-1) + fibseq(num-2)
fibseq(6)

13